crates/workspace/src/resolve.rs

```diff
@@ -0,0 +1,192 @@
+        // that the path is correctly escaped.
+        self.router.insert(path, self.settings.len() - 1).unwrap();
+    }
+
+    /// Return the appropriate [`Settings`] for a given [`Path`].
+    pub fn resolve(&self, path: &Path) -> &Settings {
+        self.router
+            .at(path.to_slash_lossy().as_ref())
+            .map(|Match { value, .. }| &self.settings[*value])
+            .unwrap_or(&self.root_settings)
+    }
+}
+
+/// Parse [Settings] from a given `air.toml`
+// TODO: Allow for an `extends` option in `air.toml`, which will make things more complex,
+// but will be very useful once we support hierarchical configuration as a way of
+// "inheriting" most top level configuration while slightly tweaking it in a nested directory.
+fn parse_settings(toml: &Path) -> Result<Settings, ParseTomlError> {
+    let options = parse_air_toml(toml)?;
+    let settings = options.into_settings();
+    Ok(settings)
+}
+
+#[derive(Debug, Error)]
+pub enum ResolvePathsError {
+    #[error(transparent)]
+    ParseToml(#[from] ParseTomlError),
+}
+
+/// For each provided `path`:
+/// - Collect the [`Settings`] associated with that `path` and store in the [`Resolver`]
+/// - Recursively search for any R files that match that `path`
+pub fn resolve_paths(
+    paths: &[PathBuf],
+) -> Result<(Vec<Result<PathBuf, ignore::Error>>, Resolver), ResolvePathsError> {
+    let paths: Vec<PathBuf> = paths.iter().map(fs::normalize_path).collect();
+
+    let mut resolver = Resolver::new();
+    let mut seen = FxHashSet::default();
+
+    // Load the `resolver` with `Settings` associated with each `path`
+    for path in &paths {
+        for ancestor in path.ancestors() {
+            if seen.insert(ancestor) {
+                if let Some(toml) = find_air_toml_in_directory(ancestor) {
+                    let settings = parse_settings(&toml)?;
+                    resolver.add(ancestor, settings);
+                    break;
+                }
+            } else {
+                // We already visited this ancestor, we can stop here.
+                break;
+            }
+        }
+    }
+
+    let (first_path, paths) = paths
+        .split_first()
+        .expect("Clap should ensure at least 1 path is supplied.");
+
+    // TODO: Parallel directory visitor
+    let mut builder = ignore::WalkBuilder::new(first_path);
+
+    for path in paths {
+        builder.add(path);
+    }
+
+    // TODO: Make these configurable options (possibly just one?)
+    // Right now we explicitly call them even though they are `true` by default
+    // to remind us to expose them.
+    //
+    // "This toggles, as a group, all the filters that are enabled by default"
+    // builder.standard_filters(true)
+    builder.hidden(true);
+    builder.parents(true);
+    builder.ignore(true);
+    builder.git_ignore(true);
+    builder.git_global(true);
+    builder.git_exclude(true);
```

Comment:
It's nice that we will get gitignore support out of the box so we don't format any directories marked in a .gitignore